File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        temp=head.next
        n=Node(head.val)
        k=n
        d={head:0}
        l=[n]
        i=1
        while(temp):
            n.next=Node(temp.val)
            d[temp]=i
            n=n.next
            l.append(n)
            temp=temp.next
            i+=1
        n=k
        l.append(None)
        d[None]=len(l)-1
        for j in range(i):
           n.random=l[d[head.random]]
           n=n.next
           head=head.next
        return k 
        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* curr=head;
        ListNode* prev=nullptr;
        int count=0;
        while(curr){
            curr=curr->next;
            count++;
        } 
        curr=head;
        while(count>=k){
            ListNode* temp=curr;
            ListNode* p=nullptr;
            ListNode* t2=curr;
            for(int i=0;i<k;i++){
                curr=curr->next;
                temp->next=p;
                p=temp;
                temp=curr;
            }if(prev)
                prev->next=p;
            else
                head=p;
            prev=t2;
            count-=k;   
        }
        if(prev)
            prev->next=curr;
        return head;
    }
};